torchtext/transforms.py

```diff
@@ -189,3 +193,350 @@ def forward(self, input: Any) -> Any:
         :rtype: Union[List[Union[str, int]], List[List[Union[str, int]]]]
         """
         return F.truncate(input, self.max_seq_len)
+
+
+class GPT2BPETokenizer(Module):
+    """
+    Transform for GPT-2 BPE Tokenizer. 
+
+    Reimplements openai GPT-2 BPE in TorchScript. Original openai implementation 
+    https://github.com/openai/gpt-2/blob/master/src/encoder.py
+
+    :param bpe_encoder: A dict mapping from bpe tokens(e.g sub-words) to
+            bpe token ids. For example: {"ingly": 4420}
+    :type bpe_encoder: Dict[str, int]
+    :param bpe_merge_ranks: A dict mapping a pair of bpe tokens to rank.
+            Torchscript doesn't support tuple as dict key type, concat
+            a pair of string with SEPERATOR (e.g \u0001) as workaround.
+            For example: {"ing\u0001ly": 210}
+    :type bpe_merge_ranks: Dict[str, int]
+    :param byte_encoder: A dict mapping a byte to an unicode character.
+    :type byte_encoder: Dict[int, str]
+    """
+    SEPERATOR: str = "\u0001"
+    LSTRIP_PATTERN: str = "\u0120"
+
+    def __init__(
+        self,
+        bpe_encoder: Dict[str, int],
+        bpe_merge_ranks: Dict[str, int],
+        byte_encoder: Dict[int, str],
+    ):
+        super().__init__()
+
+        self.bpe_encoder = bpe_encoder
+        self.bpe_merge_ranks = bpe_merge_ranks
+        self.byte_encoder = byte_encoder
+
+        self.bpe_decoder = self.load_bpe_decoder(bpe_encoder)
+        self.byte_decoder = self.load_byte_decoder(byte_encoder)
+
+        self.merge_seperator = self.SEPERATOR
+        self.lstrip_pattern = self.LSTRIP_PATTERN
+        self.inf = len(self.bpe_merge_ranks) + 1
+
+    @classmethod
+    def get_gpt2_bpe_tokenizer(
+        self,
+        encoder_json_path: str,
+        vocab_bpe_path: str,
+    ):
+        if os.path.exists(encoder_json_path):
+            encoder_json_local_path = encoder_json_path
+        else:
+            encoder_json_local_path = download_from_url(
+                url=encoder_json_path,
+                root=_CACHE_DIR
+            )
+        bpe_encoder = GPT2BPETokenizer.load_bpe_encoder(encoder_json_local_path)
+
+        if os.path.exists(vocab_bpe_path):
+            vocab_bpe_local_path = vocab_bpe_path
+        else:
+            vocab_bpe_local_path = download_from_url(
+                url=vocab_bpe_path,
+                root=_CACHE_DIR
+            )
+        bpe_merge_ranks = GPT2BPETokenizer.load_bpe_vocab(vocab_bpe_local_path)
+        return GPT2BPETokenizer(
+            bpe_encoder,
+            bpe_merge_ranks,
+            GPT2BPETokenizer.load_byte_encoder()
+        )
+
+    @classmethod
+    def load_bpe_encoder(cls, bpe_encoder_path: str) -> Dict[str, int]:
```

Comment:
Any reason to make this and the methods below `@classmethod` instead of just private methods? Do we expect the user to want to use these methods outside the context of creating a `GPT2BPETokenizer` object?

Reply:
That's a fair point. Let me refactor this one. For now, I was using the internal implementation as it is with limited modifications.

Reply:
Done! Once we refactor for Tokenizers components, we can take some of these functions and move them module level util functions.